### PR TITLE
A project can not depend on itself as a path dependency

### LINKED ISSUE #9518: A dependacy on `path = "."` shuld have a good error mesege
In a package `foo` a `[dev-dependencies] foo = { path = "." }` should be an error.
As there is no use case for it, and it does not do what most people expect.
There are not many people that will try this, and there are few configurations that don't error already. So I don't think we need to get too fancy with the comparison. A check for a path `"."` would be a big step foreword.

### PR DESCRIPTION
**Pull request by yerke**
Closes #9518

### CODE REVIEW

File: tests/testsuite/features2.rs [hunk 1]
```diff
@@ -696,17 +696,57 @@ fn cyclical_dev_dep() {
...
+        .with_status(101)
+        .with_stderr(
+            "\
+error: failed to parse manifest at [..]
+
+Caused by:
+  a project can not depend on itself as a path dependency
+",
+        )
+        .run();
```
Comment (yerke) [inline tests/testsuite/features2.rs:709]: Pretty sure this is not what this test was designed to check, but I have no idea on how to fix it properly.

**Comment by alexcrichton** [comment]
If we want to explicitly disallow this then I think it's best to do it in the cyclic detection of the resolver rather than in the manifest parsing code, since the resolver has a better idea of what connects to what.

**Comment by ehuss** [comment]
Ping @yerke Just checking in to see if you are still interested in working on this, or if you had any questions.

**Comment by yerke** [comment]
@ehuss thanks for checking in. I am sorry for taking forever. I hope I will have time around Christmas time for this.

File: tests/testsuite/features2.rs [hunk 2]
```diff
@@ -695,18 +695,14 @@ fn cyclical_dev_dep() {
...
-    p.cargo("test").run();
-
-    // New behavior.
-    switch_to_resolver_2(&p);
-    // Should decouple main.
-    p.cargo("run false").run();
-
-    // And this should be no different.
-    p.cargo("test").run();
+    // TODO: Figure out how to properly fix this test case.
```
Comment (yerke) [inline tests/testsuite/features2.rs:698]: I am not sure how to correctly test this after my changes.

**Comment by yerke** [comment]
@ehuss @alexcrichton Do you mind having a look when you get a chance? I want to know if I am on the right track. Thanks!

**Review by alexcrichton — COMMENTED** [review]
This seems to be along the right lines, yeah. Can you add a comment in `check_cycles` about what this is doing? It feels a bit odd that it's only checking for one-level-deep cycles when cycles can generally be many layers deep, but this is only applicable for `dev-dependecies`-on-itself and other packages transitively depending back on the first seems ok

**Comment by daxpedda** [comment]
There is a discussion on why this shouldn't be made a hard-error just yet here:
https://github.com/rust-lang/cargo/issues/9518#issuecomment-898384411

**Comment by yerke** [comment]
@alexcrichton Thanks for your feedback. I added the comment.

File: tests/testsuite/features2.rs (hunk 2 above)
Comment (yerke) [reply tests/testsuite/features2.rs:698]: @ehuss @alexcrichton Do you have any suggestions on how to modify that test? Thanks.

**Comment by ehuss** [comment]
I'm going to close this for now, as there is some uncertainty if this is the right thing to do (per https://github.com/rust-lang/cargo/issues/9518#issuecomment-898384411) and this is not a high priority for us to fix at this time.  Thank you for the PR, though!  We may resurrect this in the future when we have more capacity to review and clarity on how this should work.

**Pull request closed without merging**